skel/opt/cartesi/bin/init

```diff
@@ -1,85 +1,80 @@
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
-#
-
-export PATH=/opt/cartesi/bin:"$PATH"
-
-[ ! -f /opt/cartesi/etc/motd -o "$splash" == "no" ] || busybox cat /opt/cartesi/etc/motd
-
 # mount
 busybox mkdir -p /dev/pts /dev/shm
-busybox mount -t proc proc /proc -o nosuid,nodev,noexec
-busybox mount -t sysfs sys /sys -o nosuid,nodev,noexec
-busybox mount -t devpts devpts /dev/pts -o nosuid,noexec,gid=5,mode=620
-busybox mount -t tmpfs tmpfs /dev/shm -o nosuid,nodev,mode=1777
-#busybox mdev -s
+busybox mount -o nosuid,nodev,noexec          -t proc proc /proc
+busybox mount -o nosuid,noexec,mode=620,gid=5 -t devpts devpts /dev/pts
+busybox mount -o nosuid,nodev,mode=1777       -t tmpfs tmpfs /dev/shm
+busybox mount -o nosuid,nodev,noexec          -t sysfs sys /sys
+[ -d /tmp ] && busybox mount -o nosuid,nodev,mode=1777 -t tmpfs tmpfs /tmp
+[ -d /run ] && busybox mount -o nosuid,nodev,mode=0755 -t tmpfs tmpfs /run
 
-# rand
-if [ -f /opt/cartesi/var/run/random-seed ]; then
-    rndaddentropy < /opt/cartesi/var/run/random-seed
-    busybox chmod 600 /opt/cartesi/var/run/random-seed
-fi
-
-# disk
-if [ -d /mnt ]; then
-    (cd /sys/block && for DEV in *; do
-        [ ! "$DEV" = "mtdblock0" ] && \
-        NAME=$(busybox cat /sys/block/"$DEV"/device/name) && \
-        busybox mkdir "/mnt/$NAME" && \
-        busybox mount "/dev/$DEV" "/mnt/$NAME"
-    done)
-fi
+# system config
+[ -f /etc/sysctl.conf ] && busybox sysctl -pq
 
 # net
 [ -f /etc/hostname ] && busybox hostname -F /etc/hostname
-busybox ifconfig lo 127.0.0.1
+[ -z "$noloopback" ] && busybox ifconfig lo up
 
-# cmdline application
-if [ -n "$*" ]; then
-    [ -f /etc/environment ] && \
-        source /etc/environment && \
-        export PATH=/opt/cartesi/bin:"$PATH" # put it back on PATH
+# source environment
+[ -f /etc/environment ] && . /etc/environment
+export PATH="$PATH:/opt/cartesi/bin"
 
-    # can login as dapp user?
-    if [ ! "$single" == "yes" ] && busybox id dapp &> /dev/null; then
-        if [ -c /dev/rollup ]; then
-            busybox chown root:dapp /dev/rollup
-            busybox chmod 660 /dev/rollup
-        fi
-        if [ -c /dev/yield ]; then
-            busybox chown root:dapp /dev/yield
-            busybox chmod 660 /dev/yield
-        fi
-        export HOME=~dapp USER=dapp LOGNAME=dapp
-    else
-        export HOME=~root USER=root LOGNAME=root
-    fi
+# execute init from device tree when available
+[ -f /proc/device-tree/cartesi-machine/init ] && . /proc/device-tree/cartesi-machine/init
 
-    # execute cmdline
-    cd $HOME &&
+# use entrypoint from device tree when available, otherwise from command line
+if [ -s /proc/device-tree/cartesi-machine/entrypoint ]; then
+    ENTRYPOINT=$(busybox cat /proc/device-tree/cartesi-machine/entrypoint)
+elif [ -n "$*" ]; then
+    ENTRYPOINT="$*"
```

Comment:
Do we even want to support this? I think we should drop support. It's very brittle anyway.

Reply:
I do want to support this, so we can still have a way to test the same rootfs with QEMU or other RISC-V emulators. I reworked init to work in circumstances running outside our emulator, and even without emulator-tools installed.

Reply:
Alright.